### ISSUE TITLE
Option to accelerate zooming (so faster scrolls/pinches zoom faster)

### ISSUE THREAD

**Issue by bdrichards** [openseadragon/openseadragon#1860]
I honestly can't remember where I saw this idea, thought it was in this project, but must have been in another panoramic-type viewer, as I have searched here;
Is there a way to increase the rate of zoom (at an increasing rate) as the scroll speed increases?
i.e., have zoomPerSecond decrease dynamically based on speed of mousewheel scrolling...

**Comment by msalsbery** [comment]
Not built in, but one solution is to use the [OpenSeadragonViewerInput plugin](https://github.com/openseadragon-imaging/openseadragon-viewerinputhook) to intercept the wheel events then process the events any way you’d like.

Even if you don’t use the plugin, the plugin source code shows a way to access the OpenSeadragon.Viewer’s MouseTracker to intercept events before they’re handled

**Comment by iangilman** [comment]
There's also zoomPerScroll; perhaps that's what you're looking for?

**Comment by bdrichards** [comment]
Not really, the zoomPerScroll amount is the same no matter how fast a user "spins the wheel". The effect on Gigapan.com is that a faster "spin of the mouse wheel" gives you more zoomPerScroll so the zoom-in  is accelerated and you get to your desired zoom level quicker. The current OSD zoomPerScroll requires that I really pump that finger on the mousewheel because it is linear in its response. Another example, the Pinch-to-zoom response on my Surface Book responds at a faster rate (it's not linear), if I pinch faster. It is a desired effect that makes for a smoother experience when zooming-in-and-out significant amounts on a single image.

**Comment by iangilman** [comment]
Good point. This certainly seems like something worth adding. I think the big challenge will be making it work consistently across browsers, operating systems, and devices. Here's an issue for that topic: https://github.com/openseadragon/openseadragon/issues/1820

**Comment by msalsbery** [comment]
> Good point. This certainly seems like something worth adding. I think the big challenge will be making it work consistently across browsers, operating systems, and devices. Here's an issue for that topic: #1820

@iangilman Indeed! We already track pointer velocity for the flick gesture, so similar tracking could be done on wheel events. But the lack of a specification for wheel event rates is an issue as it always has been.

Definitely worth adding though! I'm chin-deep in a big MouseTracker patch right now, and this won't be included ;) but I'll give it a shot if nobody else does.

@bdrichards I confirmed the MouseTracker.scrollHandler gets raw wheel events (thankfully the zoomPerScroll implementation was done on Viewer!) so my original comment here still stands: You can intercept the viewer's scrollHandler and process the events yourself without having to modify OpenSeadragon before this feature is added

**Comment by bdrichards** [comment]
Good points.
I am eyeball-deep in getting Fall Semester virtual fieldtrips viewable with OSD, or I would take a look at your suggestion @msalsbery .
HOWEVER!!! Lookey here, I notice a year-old plugin called "OpenSeadragonSmartScrollZoom" -  'detects how fast the user is scrolling the mouse wheel and increases the zoom speed accordingly.'
In looking at this plugin, the options included seem to fit exactly my enhancement suggestion - soooooo, I will give it a try for now!

**Comment by msalsbery** [comment]
> HOWEVER!!! Lookey here, I notice a year-old plugin called "OpenSeadragonSmartScrollZoom" - 'detects how fast the user is scrolling the mouse wheel and increases the zoom speed accordingly.'

@bdrichards Nice find! Have you tried it yet?

**Comment by bdrichards** [comment]
Yes, I gave it a spin (pun intended). The parameters are tricky to get set correctly per an expected behavior: the "jump" to the accelerated logic requires a minimum # of scrolls within a time threshold. However, what is the definition of a "scroll"? Is it a "click" on my mouse wheel, or ANY movement of the wheel, no matter how far it was spun?
Still, it is a tart, for now!

**Comment by bdrichards** [comment]
oops, didn't mean to select "close".... my bad

**Comment by Deprecator16** [comment]
@bdrichards The definition of a scroll in the plugin is one "click" of the mouse wheel. Essentially, it records every scroll event that occurs using the "canvas-scroll" Viewer event and records the time between them. If the minimum # of scrolls is done within the time threshold, then every scroll after that will zoom in further, increasing linearly. Feel free to change the code however you need.

@msalsbery I can look into adding touch screen support for my plugin using pointer velocity. I never really tested it on different browsers/OS but I'll do a bit of messing around to find how it affects my plugin. I might be able to help out with the zooming speed inconsistencies.

**Comment by msalsbery** [comment]
@Deprecator16  Thanks for joining the conversation! Anything you can add to make the gesture(s) resulting from wheel events more consistent across devices/browsers would certainly be welcome. As far as touch goes, there's nothing that needs to be different - the need for speed consistency is the same, and it's the same event we're responding to.
